### ISSUE TITLE
No feature map output with "--visualize" in detect.py

### ISSUE THREAD

**Issue by FuDong623** [ultralytics/yolov5#8355]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Detection

### Bug

I used the yolov5n6 model, and I made some modifications to the network backbone such as adding C3TR. The model can be trained normally and detected using detect.py, but there is no corresponding feature map output in the result folder using '--visualize', it is an empty folder.

### Environment

-YOLO: YOLOV5 V6.1 torch1.10.1+cu113(RTX3060,12GB)
-OS: windows10
-Python: 3.8.12

### Minimal Reproducible Example

parser.add_argument('--visualize', action='store_True', help='visualize features')
python detect.py --save-txt --save-conf --visualize --hide-labels

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@FuDong623 👋 Hello! Thanks for asking about **feature visualization**. YOLOv5 🚀 features can be visualized through all stages of the model from input to output. To visualize features from a given source run `detect.py` with the `--visualize` flag:
```bash
python detect.py --weights yolov5s.pt --source data/images/bus.jpg --visualize
```

An example [Notebook](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) visualizing [bus.jpg](https://github.com/ultralytics/yolov5/blob/master/data/images/bus.jpg) features with YOLOv5s is shown below:

<a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
<img width="1811" alt="Screenshot 2021-08-30 at 16 44 04" src="https://user-images.githubusercontent.com/26833433/131357928-663a120a-3133-4e1a-8995-c812bd732e8a.png">

All stages are visualized by default, each with its own PNG showing the first 32 feature maps **output** from that stage. You can open any PNG for a closer look. For example the first 32 feature maps of the Focus() layer output are shown in `stage0_Focus_features.png`:

![stage0_Focus_features](https://user-images.githubusercontent.com/26833433/131359874-ecc9aef8-1c9b-44f5-ac17-d6b987093554.png)

Feature maps may be customized by updating the `feature_visualization()` function in `utils/plots.py`:
https://github.com/ultralytics/yolov5/blob/bb5ebc290e5d630a081d7cbc5a9725ed8cea0a24/utils/plots.py#L403-L427

Good luck 🍀 and let us know if you have any other questions!

**Comment by FuDong623** [comment]
> @FuDong623 👋 Hello! Thanks for asking about **feature visualization**. YOLOv5 🚀 features can be visualized through all stages of the model from input to output. To visualize features from a given source run `detect.py` with the `--visualize` flag:
> 
> ```shell
> python detect.py --weights yolov5s.pt --source data/images/bus.jpg --visualize
> ```
> 
> An example [Notebook](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) visualizing [bus.jpg](https://github.com/ultralytics/yolov5/blob/master/data/images/bus.jpg) features with YOLOv5s is shown below:
> 
> [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolov5) <img alt="Screenshot 2021-08-30 at 16 44 04" width="1811" src="https://user-images.githubusercontent.com/26833433/131357928-663a120a-3133-4e1a-8995-c812bd732e8a.png">
> 
> All stages are visualized by default, each with its own PNG showing the first 32 feature maps **output** from that stage. You can open any PNG for a closer look. For example the first 32 feature maps of the Focus() layer output are shown in `stage0_Focus_features.png`:
> 
> ![stage0_Focus_features](https://user-images.githubusercontent.com/26833433/131359874-ecc9aef8-1c9b-44f5-ac17-d6b987093554.png)
> 
> Feature maps may be customized by updating the `feature_visualization()` function in `utils/plots.py`:
> 
> https://github.com/ultralytics/yolov5/blob/bb5ebc290e5d630a081d7cbc5a9725ed8cea0a24/utils/plots.py#L403-L427
> 
> Good luck 🍀 and let us know if you have any other questions!

**Comment by FuDong623** [comment]
@glenn-jocher Thank you very much for your reply! It seems that using the pre-trained weights provided by yolov5 to detect images can output the corresponding feature map, but I have trained my own dataset using yolov5n6.yaml, and when I load the trained weights during detection, I cannot output the feature map by following the same operation. This is really strange.

**Comment by FuDong623** [comment]
@glenn-jocher I'm wondering if it's because my dataset is too simple, with only 1 category in it?

**Comment by FuDong623** [comment]
@glenn-jocher Thank you very much! I've found the reason. Using the "--augment" option in detect.py cannot output the feature map, it can be output without this option.

**Comment by engrjav** [comment]
Hey
Glenn thanks for your amazing work
can you tell visualization of which stage  for yolov5m should tell me about the features focused by my model for classification?
i want to check why one model classifies better than other  and understand the features focused by both models. Should it be the last feature map of stage 23?

**Comment by glenn-jocher** [comment]
@engrjav thanks for your kind words! 🙏 The feature map visualization in YOLOv5m corresponds to the last feature map of the final stage (23). However, instead of focusing solely on the final stage, I recommend analyzing feature maps at multiple stages to gain a comprehensive understanding of the model's behavior. This can help identify the differences in feature extraction between models, aiding in your comparison of their classification capabilities. Good luck with your analysis, and feel free to reach out if you have any other questions!
